llvm/lib/CodeGen/CFIFixup.cpp

```diff
@@ -10,20 +10,25 @@
 // This pass inserts the necessary  instructions to adjust for the inconsistency
 // of the call-frame information caused by final machine basic block layout.
 // The pass relies in constraints LLVM imposes on the placement of
-// save/restore points (cf. ShrinkWrap):
-// * there is a single basic block, containing the function prologue
+// save/restore points (cf. ShrinkWrap) and has certain preconditions about
+// placement of CFI instructions:
+// * for any two CFI instructions of the function prologue one dominates
+//   and is post-dominated by the other
 // * possibly multiple epilogue blocks, where each epilogue block is
 //   complete and self-contained, i.e. CSR restore instructions (and the
 //   corresponding CFI instructions are not split across two or more blocks.
-// * prologue and epilogue blocks are outside of any loops
-// Thus, during execution, at the beginning and at the end of each basic block
-// the function can be in one of two states:
+// * CFI instructions are not contained in any loops
+// Thus, during execution, at the beginning and at the end of each basic block,
+// following the prologue, the function can be in one of two states:
 //  - "has a call frame", if the function has executed the prologue, and
 //    has not executed any epilogue
 //  - "does not have a call frame", if the function has not executed the
 //    prologue, or has executed an epilogue
 // which can be computed by a single RPO traversal.
 
+// The location of the prologue is determined by finding the first block in the
```

Comment:
This is not accurate. `PrologueBlock` is the last basic block that satisfies the condition.

Reply:
Ah, sorry. This is correct. `PrologueBlock` is the first basic block in the post-order traversal.

Reply:
Perhaps we can just use `for (MachineBasicBlock &MBB : reverse(MF)) {` ?

Reply:
Ah, good point. Even if we correctly identify the last prologue block, we won't encode correctly the CFI information if the prologue blocks are not in topological order, so we might just use `reverse`.

Reply:
Done